### PR TITLE
Preempt Watson exception handling

### PR DESCRIPTION
**Pull request by mangod9**
Fix for https://github.com/dotnet/runtime/issues/35866. Switching to GC pre-emptive mode in CLR20r3BucketParamsManager::PopulateBucketParameters

### CODE REVIEW

File: src/coreclr/src/vm/dwbucketmanager.hpp [hunk 1]
```diff
@@ -954,7 +954,7 @@ bool BaseBucketParamsManager::GetFileVersionInfoForModule(Module* pModule, USHOR
     CONTRACTL
     {
         NOTHROW;
-        GC_NOTRIGGER;
+        GC_TRIGGERS;
```
Comment (mangod9) [inline src/coreclr/src/vm/dwbucketmanager.hpp:957]: possibly leave this as `GC_NOTRIGGER`?

File: src/coreclr/src/vm/dwbucketmanager.hpp [hunk 2]
```diff
@@ -475,8 +475,8 @@ void BaseBucketParamsManager::GetAppVersion(__out_ecount(maxLength) WCHAR* targe
     CONTRACTL
     {
         NOTHROW;
-        GC_NOTRIGGER;
-        MODE_ANY;
+        GC_TRIGGERS;
```
Comment (mangod9) [inline src/coreclr/src/vm/dwbucketmanager.hpp:478]: Is `GC_TRIGGERS` appropriate here although `GCX_PREEMP()` is no longer invoked here?

File: src/coreclr/src/vm/dwbucketmanager.hpp [hunk 3]
```diff
@@ -1215,11 +1215,14 @@ void CLR20r3BucketParamsManager::PopulateBucketParameters()
...
         NOTHROW;
-        GC_NOTRIGGER;
+        GC_TRIGGERS;
         MODE_ANY;
     }
     CONTRACTL_END;
 
+    // Preempt to let GC suspend
+    GCX_PREEMP();
+
```
Comment (mangod9) [inline src/coreclr/src/vm/dwbucketmanager.hpp:1225]: does `PREEMP` require a revert or will the thread switch to `GCX_COOP` when required?

Comment (jkotas) [reply src/coreclr/src/vm/dwbucketmanager.hpp:1225]: `GCX_PREEMP` does not require revert. The macro expands into an instance of an object with destructor. The object destructor will take care of the reverting at the end of the scope.

File: src/coreclr/src/vm/excep.cpp [hunk 4]
```diff
@@ -10877,7 +10877,7 @@ void EHWatsonBucketTracker::CaptureUnhandledInfoForWatson(TypeOfReportedError to
     CONTRACTL
     {
         NOTHROW;
-        GC_NOTRIGGER;
```
Comment (jkotas) [inline src/coreclr/src/vm/excep.cpp:10880]: This is called from `SetupWatsonBucketsForEscapingPreallocatedExceptions`. Change the contract on `SetupWatsonBucketsForEscapingPreallocatedExceptions` to `GC_TRIGGERS` too?

Comment (jkotas) [reply src/coreclr/src/vm/excep.cpp:10880]: Also, `GetWatsonBucketTrackerForPreallocatedException`

Comment (mangod9) [reply src/coreclr/src/vm/excep.cpp:10880]: yup, fixed. Will also run outerloop tests -- hoping that has a case for PreallocatedExceptions.

**Review by jkotas — APPROVED** [review]

**Review by Maoni0 — APPROVED** [review]

**Comment by Maoni0** [comment]
thanks for fixing this!

File: src/coreclr/src/vm/dwbucketmanager.hpp [hunk 5]
```diff
@@ -954,7 +954,7 @@ bool BaseBucketParamsManager::GetFileVersionInfoForModule(Module* pModule, USHOR
     CONTRACTL
     {
         NOTHROW;
-        GC_NOTRIGGER;
+        GC_TRIGGERS;
         MODE_ANY;
```
Comment (jkotas) [inline src/coreclr/src/vm/dwbucketmanager.hpp:958]: Should this be `MODE_PREEMPTIVE` ?

File: src/coreclr/src/vm/dwbucketmanager.hpp (hunk 1 above)
Comment (jkotas) [reply src/coreclr/src/vm/dwbucketmanager.hpp:957]: `GC_TRIGGERS` vs. `GC_NOTRIGGER` does not really matter in combination with `MODE_PREEMPTIVE`.

File: src/coreclr/src/vm/dwbucketmanager.hpp (hunk 2 above)
Comment (janvorli) [reply src/coreclr/src/vm/dwbucketmanager.hpp:478]: There doesn't seem to be anything in this method that could trigger a GC.

**Pull request merged**